src/base/notifications/notificationrequest.cpp

```diff
@@ -0,0 +1,165 @@
+#include "base/bittorrent/torrenthandle.h"
+#include "notificationsmanager.h"
+#include "notifier.h"
+
+Notifications::Request::Request()
+    : m_title(QObject::tr("Notification"))
+    , m_message(QObject::tr("Notification from qBittorrent"))
+    , m_category(Category::Generic)
+    , m_urgency(Urgency::Normal)
+    , m_severity(Severity::No)
+    , m_torrentHash(BitTorrent::InfoHash())
+    , m_timeout(-1)
+{
+}
+
+QString Notifications::Request::title() const
+{
+    return m_title;
+}
+
+QString Notifications::Request::message() const
+{
+    return m_message;
+}
+
+Notifications::Category Notifications::Request::category() const
+{
+    return m_category;
+}
+
+Notifications::Urgency Notifications::Request::urgency() const
+{
+    return m_urgency;
+}
+
+Notifications::Severity Notifications::Request::severity() const
+{
+    return m_severity;
+}
+
+BitTorrent::InfoHash Notifications::Request::torrent() const
+{
+    return m_torrentHash;
+}
+
+const Notifications::Request::ActionsMap &Notifications::Request::actions() const
+{
+    return m_actions;
+}
+
+int Notifications::Request::timeout() const
+{
+    return m_timeout;
+}
+
+const QVariant &Notifications::Request::userData() const
+{
+    return m_userData;
+}
+
+Notifications::Request::this_type &Notifications::Request::title(const QString &title)
+{
+    m_title = title;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::message(const QString &message)
+{
+    m_message = message;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::category(Category category)
+{
+    m_category = category;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::urgency(Urgency urgency)
+{
+    m_urgency = urgency;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::severity(Severity severity)
+{
+    m_severity = severity;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::torrent(BitTorrent::InfoHash &infoHash)
+{
+    m_torrentHash = infoHash;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::torrent(BitTorrent::TorrentHandle *torrent)
+{
+    m_torrentHash = torrent ? torrent->info().hash() : BitTorrent::InfoHash();
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::timeout(int timeout)
+{
+    m_timeout = timeout;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::userData(const QVariant &data)
+{
+    m_userData = data;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::addAction(const QString &id, const QString &label)
+{
+    if (!m_actions.contains(id))
+        m_actions[id] = label;
+    else
+        qDebug() << "Action with id " << id << " already exists. Skipping.";
+    return *this;
+}
+
+void Notifications::Request::exec() const &
+{
+    Manager::instance()->notifier()->showNotification(*this);
```

Comment:
Some feature implementation (I mean either some single class or some class and its satellites) shouldn't depend on how it is used. It may have many instances or be a singleton.
In your case Request is internally purposed class (implementation detail of notifications feature). It should be initialized with some Manager instance at creation time.